README.md
# Color Hash

Generate color based on the given string.

## Demo

https://zenozeng.github.io/color-hash/demo/

## Usage

### Deno

```typescript
import ColorHash from "https://deno.land/x/color_hash@v2.0.0/mod.ts"
```

### Node.js

```bash
npm install color-hash
```

```typescript
import ColorHash from 'color-hash'
```

or

```javascript
const ColorHash = require('color-hash').default;
```

### <script type="module">

```html
<script type="module">
import ColorHash from '../dist/esm.js';
</script>
```

#### Basic

```javascript
var colorHash = new ColorHash();

// in HSL, Hue ∈ [0, 360), Saturation ∈ [0, 1], Lightness ∈ [0, 1]
colorHash.hsl('Hello World'); // [ 225, 0.65, 0.35 ]

// in RGB, R, G, B ∈ [0, 255]
colorHash.rgb('Hello World'); // [ 135, 150, 197 ]

// in HEX
colorHash.hex('Hello World'); // '#8796c5'
```

#### Custom Hash Function

```javascript
var customHash = function(str) {
    var hash = 0;
    for(var i = 0; i < str.length; i++) {
        hash += str.charCodeAt(i);
    }
    return hash;
};
var colorHash = new ColorHash({hash: customHash});
colorHash.hsl('Hello World!');
colorHash.rgb('Hello World!');
colorHash.hex('Hello World!');
```

#### Custom Hue

```javascript
var colorHash = new ColorHash({hue: 90});
```

```javascript
var colorHash = new ColorHash({hue: {min: 90, max: 270}});
```

```javascript
var colorHash = new ColorHash({hue: [ {min: 30, max: 90}, {min: 180, max: 210}, {min: 270, max: 285} ]});
```

#### Custom Lightness

```javascript
var colorHash = new ColorHash({lightness: 0.5});
```

```javascript
var colorHash = new ColorHash({lightness: [0.35, 0.5, 0.65]});
```

#### Custom Saturation

```javascript
var colorHash = new ColorHash({saturation: 0.5});
```

```javascript
var colorHash = new ColorHash({saturation: [0.35, 0.5, 0.65]});
```

## License

MIT.

## FAQ

### How does it work?

It uses the `hash` function (default is BKDRHash) to calculate the hash of the given string.

```
Hue = hash % 359. (Note that 359 is a prime)
Saturation = SaturationArray[hash / 360 % SaturationArray.length]
Lightness = LightnessArray[hash / 360 / SaturationArray.length % LightnessArray.length]

By default,
SaturationArray = LightnessArray = [0.35, 0.5, 0.65]
```

### Why not LAB?

Though LAB is more perceptually uniform, HSL is easier to control.
Simply sets lightness and saturation and change hue uniformly can generate uniform colors.

## Dev

### Test

```bash
deno test
```

#### Coverage Report

https://zenozeng.github.io/color-hash/coverage/lcov-report/lib/index.html


### Follow Semantic Versioning

http://semver.org/lang/zh-CN/


CHANGELOG.md
# CHANGELOG

## v2.0.2

- Expose ESM in package.json

## v2.0.1

- chore: dist/color-hash.js: tsc --module commonjs

## v2.0.0

Note: default options were changed in v2.x. Use the following options to get the same result of v1.x
```javascript
const colorHash = new ColorHash({hash: 'bkdr', saturation: [0.65, 0.35, 0.5], lightness: [0.65, 0.35, 0.5]})
```

- refactor: rewrite using deno
- feat: New default hash function: SHA256
- feat: Deno Module: https://deno.land/x/color_hash
- feat: ESM Module: dist/esm.js
- feat: UMD Module: dist/color-hash.js

mod.ts
import BKDRHash from './lib/bkdr-hash.ts';
import { Sha256ToInt } from './lib/sha256.ts';
import { RGB2HEX, HSL2RGB } from './lib/colors.ts';

class ColorHash {

    L: number[];
    S: number[];
    hueRanges: {max: number, min: number}[]
    hash: (str: string) => number;

    constructor(options: {
        lightness?: number | number[],
        saturation?: number | number[],
        hue?: number | {max: number, min: number} | {max: number, min: number}[],
        hash?: string | ((str: string) => number);
    } = {}) {
        const [L, S] = [options.lightness, options.saturation].map(function(param) {
            param = param !== undefined ? param : [0.35, 0.5, 0.65]; // note that 3 is a prime
            return Array.isArray(param) ? param.concat() : [param];
        });
    
        this.L = L;
        this.S = S;
    
        if (typeof options.hue === 'number') {
            options.hue = {min: options.hue, max: options.hue};
        }
        if (typeof options.hue === 'object' && !Array.isArray(options.hue)) {
            options.hue = [options.hue];
        }
        if (typeof options.hue === 'undefined') {
            options.hue = [];
        }
        this.hueRanges = options.hue.map(function (range) {
            return {
                min: typeof range.min === 'undefined' ? 0 : range.min,
                max: typeof range.max === 'undefined' ? 360: range.max
            };
        });
    
        this.hash = Sha256ToInt; // Default hash function
        if (typeof options.hash === 'function') {
            this.hash = options.hash;
        } 
        if (options.hash === 'bkdr') {
            this.hash = BKDRHash;
        }
    }

    /**
     * Returns the hash in [h, s, l].
     * Note that H ∈ [0, 360); S ∈ [0, 1]; L ∈ [0, 1];
     *
     * @param {String} str string to hash
     * @returns {Array} [h, s, l]
     */
    hsl(str: string): [number, number, number] {
        var H, S, L;
        var hash = this.hash(str);
        var hueResolution = 727; // note that 727 is a prime
    
        if (this.hueRanges.length) {
            const range = this.hueRanges[hash % this.hueRanges.length];
            H = ((hash / this.hueRanges.length) % hueResolution) * (range.max - range.min) / hueResolution + range.min;
        } else {
            H = hash % 359; // note that 359 is a prime
        }
        hash = Math.ceil(hash / 360);
        S = this.S[hash % this.S.length];
        hash = Math.ceil(hash / this.S.length);
        L = this.L[hash % this.L.length];
    
        return [H, S, L];
    }


    /**
     * Returns the hash in [r, g, b].
     * Note that R, G, B ∈ [0, 255]
     *
     * @param {String} str string to hash
     * @returns {Array} [r, g, b]
     */
    rgb(str: string) {
        var hsl = this.hsl(str);
        return HSL2RGB.apply(this, hsl);
    }

    /**
     * Returns the hash in hex
     *
     * @param {String} str string to hash
     * @returns {String} hex with #
     */
    hex(str: string) {
        var rgb = this.rgb(str);
        return RGB2HEX(rgb);
    }

}



export default ColorHash;
